In [ ]:
import pandas as pd
import numpy as np

In [ ]:
#Load Train Dataset
data_train = pd.read_csv("train_dataset.csv")
data_train

Absolute Magnitude  Est Dia in KM(min)  Est Dia in KM(max)  \
0                 26.600            0.012722            0.028447   
1                 22.700            0.076658            0.171412   
2                 24.000            0.042126            0.094198   
3                 17.700            0.766576            1.714115   
4                 21.300            0.146068            0.326618   
...                  ...                 ...                 ...   
3275              23.900            0.044112            0.098637   
3276              27.300            0.009216            0.020608   
3277              21.800            0.116026            0.259442   
3278              24.786            0.029333            0.065590   
3279              21.700            0.121494            0.271669   

      Est Dia in M(min)  Est Dia in M(max)  Est Dia in Miles(min)  \
0             12.721988          28.447230               0.007905   
1             76.657557         171.411509               0.047633   
2             42.126461          94.197631               0.026176   
3            766.575574        1714.115092               0.476328   
4            146.067964         326.617897               0.090762   
...                 ...                ...                    ...   
3275          44.111820          98.637028               0.027410   
3276           9.216265          20.608196               0.005727   
3277         116.025908         259.441818               0.072095   
3278          29.332898          65.590355               0.018227   
3279         121.494041         271.668934               0.075493   

      Est Dia in Miles(max)  Est Dia in Feet(min)  Est Dia in Feet(max)  \
0                  0.017676             41.738807             93.330809   
1                  0.106510            251.501180            562.373736   
2                  0.058532            138.210178            309.047354   
3                  1.065101           2515.011805           5623.737359   
4                  0.202951            479.225620           1071.581063   
...                     ...                   ...                   ...   
3275               0.061290            144.723824            323.612307   
3276               0.012805             30.237092             67.612194   
3277               0.161210            380.662441            851.187094   
3278               0.040756             96.236546            215.191459   
3279               0.168807            398.602509            891.302306   

      Relative Velocity km per sec  ...  Inclination  Asc Node Longitude  \
0                         7.648194  ...     7.583001          347.475804   
1                        19.142824  ...     4.710039          176.203926   
2                         7.819970  ...    11.609153          146.961250   
3                         2.335267  ...     7.012588          354.475923   
4                        30.994787  ...    60.900589          219.323712   
...                            ...  ...          ...                 ...   
3275                     24.155625  ...     5.855362          232.155178   
3276                     14.509560  ...    13.442772          304.342483   
3277                      1.956978  ...     3.533539            0.907558   
3278                     22.041172  ...     9.048140          340.230154   
3279                     21.490725  ...    39.552965           51.236021   

      Orbital Period  Perihelion Distance  Perihelion Arg  Aphelion Dist  \
0         443.098049             0.890218      245.671594       1.384692   
1        1107.448956             0.984464      175.628191       3.205210   
2         317.222445             0.828952      195.589585       0.991616   
3        1036.204307             1.259659      105.208870       2.748343   
4        2174.260704             1.277179      218.504766       5.291914   
...              ...                  ...             ...            ...   
3275      705.

# Explanatory Data Analysis

In [ ]:
data_train.shape

(3280, 33)

In [ ]:
data_train.describe()

Absolute Magnitude  Est Dia in KM(min)  Est Dia in KM(max)  \
count         3280.000000         3280.000000         3280.000000   
mean            22.287687            0.197258            0.441082   
std              2.874909            0.276904            0.619175   
min             14.400000            0.001011            0.002260   
25%             20.200000            0.033462            0.074824   
50%             21.900000            0.110804            0.247765   
75%             24.500000            0.242412            0.542051   
max             32.100000            3.503926            7.835018   

       Est Dia in M(min)  Est Dia in M(max)  Est Dia in Miles(min)  \
count        3280.000000        3280.000000            3280.000000   
mean          197.257811         441.081875               0.122570   
std           276.903528         619.175112               0.172060   
min             1.010543           2.259644               0.000628   
25%            33.462237          74.823838               0.020792   
50%           110.803882         247.765013               0.068850   
75%           242.412481         542.050786               0.150628   
max          3503.926411        7835.017643               2.177238   

       Est Dia in Miles(max)  Est Dia in Feet(min)  Est Dia in Feet(max)  \
count            3280.000000           3280.000000           3280.000000   
mean                0.274075            647.171318           1447.119060   
std                 0.384737            908.476171           2031.414475   
min                 0.001404              3.315431              7.413530   
25%                 0.046493            109.784247            245.485039   
50%                 0.153954            363.529809            812.877364   
75%                 0.336815            795.316564           1778.381902   
max                 4.868453          11495.821927          25705.439285   

       Relative Velocity km per sec  ...  Inclination  Asc Node Longitude  \
count                   3280.000000  ...  3280.000000         3280.000000   
mean                      13.920372  ...    13.473183          171.349646   
std                        7.328379  ...    10.946805          103.435107   
min                        0.335504  ...     0.020987            0.001941   
25%                        8.422339  ...     5.058736           81.585368   
50%                       12.888534  ...    10.389875          172.451944   
75%                       17.972776  ...    19.762841          252.790804   
max                       44.633747  ...    75.406667          359.905890   

       Orbital Period  Perihelion Distance  Perihelion Arg  Aphelion Dist  \
count     3280.000000          3280.000000     3280.000000    3280.000000   
mean       638.009022             0.818517      184.229431       1.990773   
std        367.985549             0.240366      103.201339       0.949144   
min        176.557161             0.080744        0.006918       0.803765   
25%        369.144783             0.640433       96.476530       1.273077   
50%        508.070867             0.840137      190.594667       1.620061   
75%        789.724437             1.000085      272.468434       2.447108   
max       2556.242142             1.299832      359.929822       6.104579   

       Perihelion Time  Mean Anomaly  Mean Motion    Hazardous  
count     3.280000e+03   3280.000000  3280.000000  3280.000000  
mean      2.457733e+06    182.515907     0.731585     0.160976  
std       9.234603e+02    107.526383     0.335232     0.367564  
min       2.450854e+06      0.003191     0.140832     0.000000  
25%       2.457813e+06     87.305972     0.455855     0.000000  
50%       2.457975e+06    188.482062     0.708563     0.000000  
75%       2.458110e+06    276.559414     0.975227     0.000000  
max       2.458822e+06    359.917991     2.039000     1.000000  

[8 rows x 33 columns]

In [ ]:
data_train.isnull().sum()

Absolute Magnitude              0
Est Dia in KM(min)              0
Est Dia in KM(max)              0
Est Dia in M(min)               0
Est Dia in M(max)               0
Est Dia in Miles(min)           0
Est Dia in Miles(max)           0
Est Dia in Feet(min)            0
Est Dia in Feet(max)            0
Relative Velocity km per sec    0
Relative Velocity km per hr     0
Miles per hour                  0
Miss Dist.(Astronomical)        0
Miss Dist.(lunar)               0
Miss Dist.(kilometers)          0
Miss Dist.(miles)               0
Orbit ID                        0
Orbit Uncertainity              0
Minimum Orbit Intersection      0
Jupiter Tisserand Invariant     0
Epoch Osculation                0
Eccentricity                    0
Semi Major Axis                 0
Inclination                     0
Asc Node Longitude              0
Orbital Period                  0
Perihelion Distance             0
Perihelion Arg                  0
Aphelion Dist                   0
Perihelion Tim

The dataset have shape (3280, 33). It means the data have 3280 data with 33 features. There is not any missing values. 

In [ ]:
#Checking the correlation od various columns to the target columns
corr_matrix = data_train.corr()
corr_matrix['Hazardous'].sort_values(ascending=True)

Orbit Uncertainity             -0.326615
Absolute Magnitude             -0.326416
Minimum Orbit Intersection     -0.292193
Perihelion Distance            -0.202460
Jupiter Tisserand Invariant    -0.017197
Perihelion Arg                 -0.006480
Mean Motion                    -0.000756
Orbital Period                  0.000625
Semi Major Axis                 0.001562
Inclination                     0.005087
Asc Node Longitude              0.020257
Miss Dist.(Astronomical)        0.032267
Miss Dist.(miles)               0.032267
Miss Dist.(lunar)               0.032267
Miss Dist.(kilometers)          0.032267
Perihelion Time                 0.035654
Epoch Osculation                0.039864
Mean Anomaly                    0.046276
Aphelion Dist                   0.052986
Est Dia in Miles(min)           0.174192
Est Dia in KM(min)              0.174192
Est Dia in KM(max)              0.174192
Est Dia in Feet(min)            0.174192
Est Dia in Feet(max)            0.174192
Est Dia in M(max

There are seven columns that weaker that the others features.

# Separating Input Features and Output Features

In [ ]:
X = data_train.drop("Hazardous",axis = 1)
y = data_train.Hazardous

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [ ]:
from sklearn.preprocessing import LabelEncoder, StandardScaler
scaler = StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

# Predicting Model

In [ ]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, accuracy_score, f1_score, precision_recall_curve, roc_auc_score, roc_curve

## Using Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
reg = LogisticRegression()
reg.fit(X_train, y_train)
y_pred_1 = reg.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score
print('Accuracy score',accuracy_score(y_pred_1,y_test))

Accuracy score 0.9502032520325203


## Using SGD Classifier
This estimator implements regularized linear models with stochastic gradient descent (SGD) learning: the gradient of the loss is estimated each sample at a time and the model is updated along the way with a decreasing strength schedule (aka learning rate)

In [ ]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, y_train)
y_pred_2 = sgd.predict(X_test)

In [ ]:
print('Recall Score',recall_score(y_pred_2,y_test))
print('Precision Score', precision_score(y_pred_2,y_test))
print('Accuracy Score', accuracy_score(y_pred_2,y_test))

Recall Score 0.8728323699421965
Precision Score 0.8342541436464088
Accuracy Score 0.9471544715447154


## Using Random Forest Classifier

In [ ]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.ensemble import RandomForestClassifier

forest_clf = RandomForestClassifier(n_estimators=100, random_state=42)
forest_clf.fit(X_train, y_train)
y_probas_forest = cross_val_predict(forest_clf, X_train, y_train, cv=3,method="predict_proba")

In [ ]:
y_train_pred_forest = cross_val_predict(forest_clf, X_train, y_train, cv=3)
precision_score(y_train, y_train_pred_forest)

1.0

In [ ]:
#Confusion matrix on train dataset
confusion_matrix(y_train, y_train_pred_forest)

array([[1949,    0],
       [   8,  339]])

In [ ]:
print('Recall Score', recall_score(y_train, y_train_pred_forest))
print('Accuracy Score', accuracy_score(y_train, y_train_pred_forest))

Recall Score 0.9769452449567724
Accuracy Score 0.9965156794425087


In [ ]:
#Predicting with test set
yf_pred = forest_clf.predict(X_test)

In [ ]:
#Displaying the accuracy score for the test set
accuracy_score(y_test, yf_pred)

0.9969512195121951

## Using KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=2)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

In [ ]:
#Displaying the accuracy score for the test set
accuracy_score(y_test, y_pred)

0.8516260162601627

## Using Naive Bayes

In [ ]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(X_train, y_train)
y_pred = gnb.predict(X_test)

In [ ]:
#Displaying the accuracy score for the test set
accuracy_score(y_test, y_pred)

0.9085365853658537

## Using MLP

In [ ]:
from sklearn.neural_network import MLPRegressor

regr = MLPRegressor(random_state = 1, max_iter = 500).fit(X_train, y_train)

In [ ]:
#Prediksi data test
regr.predict(X_test[:2])
pred = regr.score(X_test, y_test)
pred

0.77169884010367

In [ ]:
#Nilai regresi pada data test
test = regr.score(X_train, y_train)
test

0.8741061057216297

In [ ]:
#Nilai akurasi 
accuracy = 1 - abs(test-pred)
print("Accuracy from the model:",accuracy*100)

Accuracy from the model: 89.75927343820402


## SVM

In [ ]:
from sklearn.svm import SVC
classifier = SVC(kernel='rbf', C=1000, gamma=0.001)
classifier.fit(X=X_train, y=y_train)

SVC(C=1000, gamma=0.001)

In [ ]:
y_pred = classifier.predict(X=X_test)
y_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0,

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))
accuracy_score(y_true=y_test, y_pred=y_pred)

[[785  18]
 [ 28 153]]
              precision    recall  f1-score   support

           0       0.97      0.98      0.97       803
           1       0.89      0.85      0.87       181

    accuracy                           0.95       984
   macro avg       0.93      0.91      0.92       984
weighted avg       0.95      0.95      0.95       984



0.9532520325203252

The best model for this data is Random forest

# Load Test Data

In [ ]:
data_test = pd.read_csv("test_dataset.csv")

In [ ]:
data_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1407 entries, 0 to 1406
Data columns (total 32 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Absolute Magnitude            1407 non-null   float64
 1   Est Dia in KM(min)            1407 non-null   float64
 2   Est Dia in KM(max)            1407 non-null   float64
 3   Est Dia in M(min)             1407 non-null   float64
 4   Est Dia in M(max)             1407 non-null   float64
 5   Est Dia in Miles(min)         1407 non-null   float64
 6   Est Dia in Miles(max)         1407 non-null   float64
 7   Est Dia in Feet(min)          1407 non-null   float64
 8   Est Dia in Feet(max)          1407 non-null   float64
 9   Relative Velocity km per sec  1407 non-null   float64
 10  Relative Velocity km per hr   1407 non-null   float64
 11  Miles per hour                1407 non-null   float64
 12  Miss Dist.(Astronomical)      1407 non-null   float64
 13  Mis

There are no missing values

In [ ]:
data_test = scaler.transform(data_test)

In [ ]:
data_target = forest_clf.predict(data_test)

In [ ]:
res = pd.DataFrame(data_target) # target are nothing but the final predictions of your model on input features of your new unseen test data
res.columns = ["Hazardous"]

# To download the csv file locally
from google.colab import files
res.to_csv('submission.csv', index = False)
files.download('submission.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>